### PR TITLE
Fix 5869: Google OAuth should work by default

### PR DESCRIPTION
**Pull request by jumde**
Fix https://github.com/brave/brave-browser/issues/5869

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

1. Navigate to expensify.com 
2. Verify that Cross-Site cookie blocking is enabled
3. Click login using Google
4. Login should succeed
5. Navigate to `brave://settings/socialBlocking` and disable Google login
6. Navigate to `brave://settings/clearBrowserData` and clear all data
7. Try logging in again, login should fail

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_shields/browser/brave_shields_util.cc [hunk 1]
```diff
@@ -205,6 +205,20 @@ ControlType GetCookieControlType(Profile* profile, const GURL& url) {
       HostContentSettingsMapFactory::GetForProfile(profile), url);
 }
 
+void SetCookieWhitelist(HostContentSettingsMap* map, ControlType type,
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_util.cc]: it's confusing to me that we have a method called `SetCookieWhitelist` that takes a ControlType. It seems to me that this is just an overloaded `SetCookieControlType` with a string pattern instead of a GURL, correct?

File: components/brave_shields/browser/brave_shields_util.cc [hunk 2]
```diff
@@ -205,6 +205,20 @@ ControlType GetCookieControlType(Profile* profile, const GURL& url) {
...
+                        std::string pattern) {
+  map->SetContentSettingCustomScope(ContentSettingsPattern::Wildcard(),
+                                    ContentSettingsPattern::FromString(pattern),
+                                    CONTENT_SETTINGS_TYPE_PLUGINS, kCookies,
+                                    GetDefaultBlockFromControlType(type));
+
+  // Allow cookies for social media sites when used in first party context
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_util.cc]: why does this method care whether it's social media or not?

File: components/brave_shields/browser/brave_shields_util.cc [hunk 3]
```diff
@@ -205,6 +205,20 @@ ControlType GetCookieControlType(Profile* profile, const GURL& url) {
...
+  map->SetContentSettingCustomScope(
+      ContentSettingsPattern::FromString(pattern),
+      ContentSettingsPattern::FromString("https://firstParty/*"),
+      CONTENT_SETTINGS_TYPE_PLUGINS, kCookies, CONTENT_SETTING_ALLOW);
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_util.cc]: why is this always `ALLOW` if we're passing a ControlType?

File: components/brave_shields/browser/cookie_pref_service.cc [hunk 4]
```diff
@@ -21,6 +23,21 @@ namespace brave_shields {
...
+  if (IntToContentSetting(prefs->GetInteger(
+          "profile.default_content_setting_values.cookies")) ==
+          ContentSetting::CONTENT_SETTING_ALLOW &&
+      prefs->GetBoolean(prefs::kBlockThirdPartyCookies)) {
+    // Only set social media exceptions when cross-site cookie blocking
+    // is enabled
+    ControlType type = prefs->GetBoolean(kGoogleLoginControlType)
+                           ? ControlType::ALLOW
+                           : ControlType::BLOCK;
+    SetCookieWhitelist(map, type, kGoogleOAuthPattern);
```
Comment (bridiver) [inline components/brave_shields/browser/cookie_pref_service.cc]: I think it's problematic to set this as a brave cookie plugin type because it will always end up with an override one direction or the other which could potentially conflict with user overrides. It seems better to convert this directly to cookie settings in BravePrefProvider so we only add an override when it's enabled

Comment (bridiver) [reply components/brave_shields/browser/cookie_pref_service.cc]: it's especially problematic because of this https://github.com/brave/brave-core/pull/3784/files#diff-6c160ccdf43d9f23e468a95ec08b4d12R219 which adds an allow override no matter what the setting is

File: components/brave_shields/browser/cookie_pref_service.cc [hunk 5]
```diff
@@ -21,6 +23,21 @@ namespace brave_shields {
 
 namespace {
 
+void SetCookieWhitelistForSocialMedia(HostContentSettingsMap* map,
```
Comment (bridiver) [inline components/brave_shields/browser/cookie_pref_service.cc]: this method makes more sense as `SetCookieWhitelist` because it will potentially add other pref based overrides, but `ForSocialMedia` seems irrelevant to the method and google auth isn't social media anyway. This is just setting a whitelist based on prefs

**Comment by bridiver** [comment]
please add tests for this including a test to ensure that no overrides are added when the pref is false

File: components/brave_shields/browser/brave_shields_util.h [hunk 6]
```diff
@@ -43,6 +43,11 @@ void SetCookieControlType(Profile* profile, ControlType type, const GURL& url);
 void SetCookieControlType(HostContentSettingsMap* map,
                           ControlType type,
                           const GURL& url);
+
+void SetCookieWhitelist(HostContentSettingsMap* map,
+                        ControlType type,
+                        std::string pattern);
```
Comment (iefremov) [inline components/brave_shields/browser/brave_shields_util.h]: `const std::string&` or `base::StringPiece`

File: components/brave_shields/browser/brave_shields_util.cc (hunk 1 above)
Comment (iefremov) [reply components/brave_shields/browser/brave_shields_util.cc]: I'd say `SetCookieControlTypeForPattern`? Overloads are often confusing and also not encouraged by CC.

File: browser/net/brave_network_delegate_browsertest.cc [hunk 7]
```diff
@@ -71,6 +89,29 @@ class BraveNetworkDelegateBrowserTest : public InProcessBrowserTest {
                                           top_level_page_url_);
   }
 
+  void BlockGoogleOAuthCookies() {
+    browser()->profile()->GetPrefs()->SetBoolean(kGoogleLoginControlType,
+                                                 false);
+  }
+
+  void NavigateToPageWithFrame(const std::string& host) {
+      GURL main_url(https_server_.GetURL(host, "/iframe.html"));
```
Comment (iefremov) [inline browser/net/brave_network_delegate_browsertest.cc]: pls correct spaces everywhere (or use `clang-format`)

File: components/brave_shields/common/brave_shield_constants.h [hunk 8]
```diff
@@ -17,6 +17,7 @@ const char kFingerprinting[] = "fingerprinting";
 const char kBraveShields[] = "braveShields";
 const char kReferrers[] = "referrers";
 const char kCookies[] = "cookies";
+const char kGoogleLogin[] = "google-login";
```
Comment (iefremov) [inline components/brave_shields/common/brave_shield_constants.h]: is it used anywhere?

File: components/content_settings/core/browser/brave_content_settings_pref_provider.h [hunk 9]
```diff
@@ -56,6 +56,8 @@ class BravePrefProvider : public PrefProvider,
                                const ContentSettingsPattern& secondary_pattern,
                                ContentSettingsType content_type,
                                const std::string& resource_identifier) override;
+  void OnCookiePrefsChanged(const std::string& pref);
+  void SetOAuthCookieException(ContentSettingsType type);
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.h]: unused?

File: browser/net/brave_network_delegate_browsertest.cc [hunk 10]
```diff
@@ -152,3 +194,23 @@ IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBrowserTest,
...
+  NavigateToPageWithFrame("a.com");
+  ExpectCookiesOnHost("accounts.google.com", "");
+
+  NavigateFrameTo("accounts.google.com", "/set-cookie?oauth=true");
+  ExpectCookiesOnHost("accounts.google.com", "oauth=true");
+}
+
+IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBrowserTest,
+                       IframeGoogleAccountsCookieBlocked) {
+  BlockGoogleOAuthCookies();
```
Comment (iefremov) [inline browser/net/brave_network_delegate_browsertest.cc]: we need a test, showing that the new preference doesn't override user choice in per-site shields settings

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 11]
```diff
@@ -254,6 +261,14 @@ void BravePrefProvider::UpdateCookieRules(ContentSettingsType content_type,
     }
   }
 
+  if (prefs_->GetBoolean(kGoogleLoginControlType)) {
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: not sure if this overrides user shields settings or not.
Anyway, this method becomes too fat, so i'd prefer to put it into a separate method (or at least add a comment).

File: browser/net/brave_network_delegate_browsertest.cc (hunk 10 above)
Comment (bridiver) [reply browser/net/brave_network_delegate_browsertest.cc]: I actually requested this in the last review and should verify your comment below

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 12]
```diff
@@ -314,6 +329,12 @@ void BravePrefProvider::NotifyChanges(const std::vector<Rule>& rules,
            "");
   }
 }
+
+void BravePrefProvider::OnCookiePrefsChanged(
+    const std::string& pref) {
+    OnCookieSettingsChanged(CONTENT_SETTINGS_TYPE_PLUGINS);
```
Comment (bridiver) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: lint - 2 spaces

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc (hunk 11 above)
Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: it shouldn't override shields settings as long as it comes last, but it should have a test to verify

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 13]
```diff
@@ -190,8 +199,40 @@ std::unique_ptr<RuleIterator> BravePrefProvider::GetRuleIterator(
                                        incognito);
 }
 
+void BravePrefProvider::AddCookieExceptions(ContentSetting content_type,
```
Comment (bridiver) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: this method should take a pattern, making a generic `AddCookieExceptions` still doesn't make sense if it is hard coded to kGoogleOAuthPattern

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 14]
```diff
@@ -254,6 +303,43 @@ void BravePrefProvider::UpdateCookieRules(ContentSettingsType content_type,
     }
   }
 
+  if (prefs_->GetBoolean(kGoogleLoginControlType)) {
+    if (prefs_->GetInteger("profile.default_content_setting_values.cookies") == CONTENT_SETTING_BLOCK) {
+      AddCookieExceptions(CONTENT_SETTING_ALLOW, incognito);
+    }
+
+    if (prefs_->GetBoolean(prefs::kBlockThirdPartyCookies))  {
```
Comment (bridiver) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: the current cookie setting is not relevant, this is an override that is independent of cookie settings

File: browser/net/brave_network_delegate_browsertest.cc (hunk 10 above)
Comment (jumde) [reply browser/net/brave_network_delegate_browsertest.cc]: Added here: https://github.com/brave/brave-core/pull/3784/files#diff-a3b1f2bffde11cef192a185a65955508R350

File: browser/net/brave_network_delegate_browsertest.cc [hunk 15]
```diff
@@ -188,7 +212,7 @@ IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBrowserTest, DefaultCookiesBlocked) {
 IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBrowserTest,
                        PrefToggleBlockAllToBlockThirdParty) {
   DefaultBlockAllCookies();
-  BlockThirdPartyCookies();
+  BlockAllThirdPartyCookies();
```
Comment (jumde) [inline browser/net/brave_network_delegate_browsertest.cc]: Renamed so that it's consistent with other helper functions like `DefaultBlockAllCookies`

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc (hunk 11 above)
Comment (jumde) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: Test added here: https://github.com/brave/brave-core/pull/3784/files#diff-a3b1f2bffde11cef192a185a65955508R350

Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: although actually we want it to override site shield settings because it should work the same whether the default is block 3p cookies or whether the default is something else and site shield settings are block 3p cookies. In the case where site settings are block all you won't be able to login to anything anyway

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 16]
```diff
@@ -193,8 +200,18 @@ std::unique_ptr<RuleIterator> BravePrefProvider::GetRuleIterator(
 void BravePrefProvider::UpdateCookieRules(ContentSettingsType content_type,
                                           bool incognito) {
   auto& rules = cookie_rules_[incognito];
+  auto old_rules = std::move(brave_cookie_rules_[incognito]);
+
   rules.clear();
 
+  if (prefs_->GetBoolean(kGoogleLoginControlType)) {
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: I desperately desire a comment here, describing the whole situation. So an occasional stranger could have a little chance to understand why and what happens here :)

File: browser/net/brave_network_delegate_browsertest.cc [hunk 17]
```diff
@@ -67,21 +82,28 @@ class BraveNetworkDelegateBrowserTest : public InProcessBrowserTest {
...
 
-  void AllowCookies() {
+  void AllowCookies(const EmbeddedTestServer* server) {
     brave_shields::SetCookieControlType(browser()->profile(),
                                         brave_shields::ControlType::ALLOW,
-                                        top_level_page_url_);
+                                        server->GetURL("a.com", "/"));
+  }
+
+  void BlockThirdPartyCookies(const EmbeddedTestServer* server) {
```
Comment (bridiver) [inline browser/net/brave_network_delegate_browsertest.cc]: I think this method name is confusing because it's not obvious in the tests that this is a site override and not the default. I think it's better to pass in the URL instead of the server

File: browser/net/brave_network_delegate_browsertest.cc [hunk 18]
```diff
@@ -67,21 +82,28 @@ class BraveNetworkDelegateBrowserTest : public InProcessBrowserTest {
         browser()->profile(), brave_shields::ControlType::ALLOW, GURL());
   }
 
-  void AllowCookies() {
+  void AllowCookies(const EmbeddedTestServer* server) {
```
Comment (bridiver) [inline browser/net/brave_network_delegate_browsertest.cc]: same as below

File: browser/net/brave_network_delegate_browsertest.cc (hunk 17 above)
Comment (bridiver) [reply browser/net/brave_network_delegate_browsertest.cc]: or pass both

File: browser/net/brave_network_delegate_browsertest.cc [hunk 19]
```diff
@@ -67,21 +82,28 @@ class BraveNetworkDelegateBrowserTest : public InProcessBrowserTest {
...
+
+  void BlockThirdPartyCookies(const EmbeddedTestServer* server) {
+    brave_shields::SetCookieControlType(
+        browser()->profile(),
+        brave_shields::ControlType::BLOCK_THIRD_PARTY,
+        server->GetURL("a.com", "/"));
   }
 
-  void BlockCookies() {
+  void BlockCookies(const EmbeddedTestServer* server) {
```
Comment (bridiver) [inline browser/net/brave_network_delegate_browsertest.cc]: same ^^

File: browser/net/brave_network_delegate_browsertest.cc [hunk 20]
```diff
@@ -51,12 +61,17 @@ class BraveNetworkDelegateBrowserTest : public InProcessBrowserTest {
...
+    command_line->AppendSwitch(switches::kIgnoreCertificateErrors);
+  }
+
   void DefaultBlockAllCookies() {
     brave_shields::SetCookieControlType(
         browser()->profile(), brave_shields::ControlType::BLOCK, GURL());
   }
 
-  void BlockThirdPartyCookies() {
+  void BlockAllThirdPartyCookies() {
```
Comment (bridiver) [inline browser/net/brave_network_delegate_browsertest.cc]: `BlockAllThirdPartyCookies` doesn't mean anything different to me than `BlockThirdPartyCookies`. `DefaultBlockThirdPartyCookies` maybe?

File: browser/net/brave_network_delegate_browsertest.cc [hunk 21]
```diff
@@ -298,3 +329,58 @@ IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBrowserTest,
   ExpectCookiesOnHost(embedded_test_server(), "a.com", "");
   ExpectCookiesOnHost(embedded_test_server(), "b.com", "");
 }
+
+IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBrowserTest,
+                       IframeGoogleAccountsCookieAllowed) {
```
Comment (bridiver) [inline browser/net/brave_network_delegate_browsertest.cc]: should these maybe be `ThirdPartyGoogleAccountsCookiesAllowed`? Instead of `Iframe`? I know it's basically equivalent, but I think `ThirdParty` is a better description

**Comment by bsclifton** [comment]
🎉 !!
